tests/pf/test_forward_Mag_Linear.py

```diff
@@ -101,6 +105,98 @@ def get_block_inds(grid, block):
     np.testing.assert_allclose(d_t, d @ tmi)
 
 
+def test_ana_mag_tmi_grad_forward():
```

Comment:
Thanks for adding this test function. I think it's nice to have the Simulation compared with the forwards in geoana.

Nevertheless, I think it would be nice to add a finite difference test: forward model the TMI of a set of prisms on points separated by a small distance, so we can then approximate the derivatives of the TMI by finite differences of these values. This way we can have a test that don't rely on the third-order derivatives of the kernels, but on the second-order ones.

What do you think?

Reply:
a bit cuncumber some if we try to keep the problem small, but I cooked up something. see what you think.

Reply:
The ones in geoana are tested as derivatives of the kernels, so I think it's sufficient to compare against geoana.

Reply:
Sorry for the misunderstanding. I wasn't saying to increase the number of receivers in the grid, but to create "shifted" grids that could have the original 5 x 5 shape. For example, define a pair of grids shifted along `x`:

```python 
    nx, ny = 5, 5
    xr = np.linspace(-20, 20, nx)
    yr = np.linspace(-20, 20, ny)
    X, Y = np.meshgrid(xr, yr)

    delta_x = 0.01
    receivers_loc = np.c_[X.reshape(-1), Y.reshape(-1), Z.reshape(-1)]
    receivers_loc_pos = np.c_[X.reshape(-1) + delta_x / 2, Y.reshape(-1), Z.reshape(-1)]
    receivers_loc_neg = np.c_[X.reshape(-1) - delta_x / 2, Y.reshape(-1), Z.reshape(-1)]
```

Then compute:
* `tmi_x` on `receivers_loc`
* `tmi` on `receivers_loc_pos` and `receivers_loc_neg`

And approximate the `tmi_x` by finite differences:
```python
tmi_x_finite_diff = (tmi_pos - tmi_neg) / delta_x
```

Reply:
> The ones in `geoana` are tested as derivatives of the kernels, so I think it's sufficient to compare against `geoana`.

I'm not saying to retest `geoana` kernels here. In the tests that @thibaut-kobold wrote, we are implementing a simplified version of the forward of the derivatives of tmi, particularly in these lines:
```python
    d = (
        prism_1.magnetic_field_gradient(locXyz)
        + prism_2.magnetic_field_gradient(locXyz)
        + prism_3.magnetic_field_gradient(locXyz)
    ) * mu_0
    tmi_x = (d[:, 0, 0] * b0[0] + d[:, 0, 1] * b0[1] + d[:, 0, 2] * b0[2]) / H0[0]
    tmi_y = (d[:, 1, 0] * b0[0] + d[:, 1, 1] * b0[1] + d[:, 1, 2] * b0[2]) / H0[0]
    tmi_z = (d[:, 2, 0] * b0[0] + d[:, 2, 1] * b0[1] + d[:, 2, 2] * b0[2]) / H0[0]
    np.testing.assert_allclose(d_x, tmi_x, rtol=1e-10, atol=1e-12)
    np.testing.assert_allclose(d_y, tmi_y, rtol=1e-10, atol=1e-12)
    np.testing.assert_allclose(d_z, tmi_z, rtol=1e-10, atol=1e-12)
```

In the event that both the implementation in the simulation and this simplified implementation have the same error, tests will pass and won't be able to catch the bug. By adding the finite difference test we can solve that.

It's not critical, it's just something it would be nice to have.

Reply:
Thanks @santisoler, I did add a finite difference test yesterday as you suggested, but with one global receiver grid rather than 5 separate grids as you propose.
I think having a single grid make the test syntax less cumbersome and easier to debug later on.
The test is still very fast to run with the increased number of receivers (<3 s on my machine).
See if that solution works for you.